### PR TITLE
Route logs back from host to subscribed clients.

### PR DESCRIPTION
**Pull request by joeldudley**
### Why is this change needed?

The enclave includes logs in the `BlockSubmissionResponse`s it sends back to the host. However, the host does not route these back to the client.

### What changes were made as part of this PR:

- Routes logs back to subscribers

Note the following caveats:

* All logs are sent to all subscribers
* Only the `eth_subscribe` endpoint is supported, and not the other methods to get logs

You can test this by:

* Starting the full network sim
* Running the following client defined in some `main.go` file:

```
package main

import (
	"context"
	"fmt"
	"github.com/ethereum/go-ethereum/rlp"
	"github.com/obscuronet/go-obscuro/go/common"
	"time"

	"github.com/ethereum/go-ethereum/core/types"

	"github.com/ethereum/go-ethereum/rpc"
)

const url = "ws://127.0.0.1:37500/" // The websocket address for the first Obscuro node in the full network simulation.

func main() {
	var client *rpc.Client
	for {
		var err error
		client, err = rpc.DialWebsocket(context.Background(), url, "")
		if err == nil {
			break
		}
		time.Sleep(time.Second)
	}
	defer client.Close()

	ch := make(chan *types.Log)

	logSubscription := common.LogSubscription{
		Accounts: []*common.SubscriptionAccount{},
	}
	encodedLogSubscription, err := rlp.EncodeToBytes(logSubscription)
	if err != nil {
		panic(err)
	}

	sub, err := client.Subscribe(context.Background(), "eth", ch, "logs", encodedLogSubscription)
	if err != nil {
		panic(err)
	}

	for {
		select {
		case log := <-ch:
			println(fmt.Sprintf("Received logs. Block number: %d. Index: %d. Data: %s", log.BlockNumber, log.Index, string(log.Data)))
		case err = <-sub.Err():
			panic(err)
		}
	}
}
```

You should then receive the logs over the created subscription.

### What are the key areas to look at

- Describe the key areas for a reviewer to look at 

### :rotating_light: Definition of done :rotating_light:
- [ ] Unit tests added to cover new or changed functionality 
- [ ] Docs pages updated to cover new or changed functionality
- [ ] [Changelog.md](https://github.com/obscuronet/go-obscuro/blob/main/docs/testnet/changelog.md) updated

### CODE REVIEW

File: go/host/node/node.go [hunk 1]
```diff
@@ -152,6 +158,12 @@ func NewHost(
 				Service:   clientapi.NewTestAPI(node),
 				Public:    true,
 			},
+			{
+				Namespace: "eth",
+				Version:   apiVersion1,
+				Service:   filters.NewPublicFilterAPI(events.NewBackend(node.logsCh), false, 5*time.Minute),
```
Comment (joeldudley) [inline go/host/node/node.go:164]: We reuse Geth's `PublicFilterAPI` to handle log subscriptions, passing in our own custom backend.

File: go/host/node/node.go [hunk 2]
```diff
@@ -305,6 +316,32 @@ func (a *Node) ReceiveTx(tx common.EncryptedTx) {
...
+	if err != nil {
+		return fmt.Errorf("could not generate new UUID for subscription")
+	}
+
+	// TODO - #453 - Link subscriptions to specific accounts.
+	subscription := common.LogSubscription{
+		Accounts: []*common.SubscriptionAccount{},
+	}
+
+	// TODO - #453 - Encrypt subscription requests, rather than just encoding them.
```
Comment (tudor-malene) [inline go/host/node/node.go:331]: the encrypted subscription should be passed in as a parameter from the rpc

File: go/host/node/node.go [hunk 3]
```diff
@@ -305,6 +316,32 @@ func (a *Node) ReceiveTx(tx common.EncryptedTx) {
...
+	encodedSubscription, err := rlp.EncodeToBytes(subscription)
+	if err != nil {
+		return fmt.Errorf("could not encrypt subscription request")
+	}
+
+	err = a.EnclaveClient().Subscribe(id, encodedSubscription)
```
Comment (tudor-malene) [inline go/host/node/node.go:337]: I would have imagined that the host would maintain a mapping from the generated subscription id to a channel where encrypted log entries would be dispatched

File: go/host/node/node.go [hunk 4]
```diff
@@ -562,6 +601,20 @@ func (a *Node) storeBlockProcessingResult(result common.BlockSubmissionResponse)
 	}
 }
 
+// Distributes logs to subscribed clients.
+// TODO - #453 - Encrypt logs, rather than just serialising them as JSON.
+// TODO - #453 - Distribute logs specifically based on subscription IDs, rather than sending all logs to everyone.
+func (a *Node) sendLogsToSubscribers(result common.BlockSubmissionResponse) {
+	for _, jsonLogs := range result.SubscribedLogs {
```
Comment (tudor-malene) [inline go/host/node/node.go:608]: I see there are todos here which I think capture the requirements.
Basically, based on the `result.SubscribedLogs` key, this has to do the dispatching of encrypted events. There should be no unmarshalling/unencryption here.

File: go/host/rpc/clientapi/client_api_filter.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+// FilterAPI exposes a subset of Geth's PublicFilterAPI operations.
+type FilterAPI struct {
+	host          host.Host
+	gethFilterAPI *filters.PublicFilterAPI
+}
+
+func NewFilterAPI(host host.Host, logsCh chan []*types.Log) *FilterAPI {
+	return &FilterAPI{
+		host:          host,
+		gethFilterAPI: filters.NewPublicFilterAPI(events.NewBackend(logsCh), false, 5*time.Minute),
```
Comment (tudor-malene) [inline go/host/rpc/clientapi/client_api_filter.go:25]: we'll have to brainstorm this a bit.

I think the "PublicFilterAPI" is something that has to be exposed only inside the enclave, and on the wallet extension (auth client).

On the host there should be something that takes in an encrypted subscription request and exposes `event.Subscription` implementations that are Obscuro specific


It's fine to use the current stuff as a stepping stone, but we have to keep this in mind.

**Review by tudor-malene — COMMENTED** [review]
@joeldudley, let's have a chat when you're free.

File: go/host/events/backend.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+	logsCh <-chan []*types.Log
+}
+
+func NewBackend(logsCh <-chan []*types.Log) Backend {
+	return Backend{
+		logsCh: logsCh,
+	}
+}
+
+func (b Backend) ChainDb() ethdb.Database { //nolint:stylecheck
```
[the remark this answers is not in the document]
Comment (joeldudley) [reply go/host/events/backend.go:41]: Not sure what you mean exactly. But this is the interface method, I don't have control of this.

File: go/host/node/node.go [hunk 7]
```diff
@@ -152,6 +157,12 @@ func NewHost(
 				Service:   clientapi.NewTestAPI(node),
 				Public:    true,
 			},
+			{
+				Namespace: "eth",
```
[the remark this answers is not in the document]
Comment (joeldudley) [reply go/host/node/node.go:161]: Fixed, good catch.

File: go/host/node/node.go (hunk 2 above)
Comment (joeldudley) [reply go/host/node/node.go]: Added.

**Review by tudor-malene — APPROVED** [review]
lgtm

**Pull request merged**